### PR TITLE
[FLINK-2161] modified Scala shell start script to take additional argument (-a or --addclasspath…

### PR DESCRIPTION
**Pull request by nikste**
…) for external libraries

### CODE REVIEW

**Comment by aljoscha** [comment]
I think this is only half of what is needed. The other half would be sending the jar files along when submitting the job with the modified RemoteEnvironment.

**Comment by nikste** [comment]
You are right. I only tried it with the local execution environment. It does not work if you start a separate job manager on your machine.

**Comment by nikste** [comment]
This should work now. It sends the jar alongside the compiled user-code to the cluster.

File: flink-staging/flink-scala-shell/src/main/scala/org.apache.flink/api/scala/FlinkILoop.scala [hunk 1]
```diff
@@ -191,5 +203,14 @@ HINT: You can use print() on a DataSet to print the contents to this shell.
...
 
   }
+  
+  def getExternalJars(): Array[String] ={
+    var extJars : Array[String] = Array.empty[String]
+    this.externalJars match{
+      case Some(ej) => extJars = ej
+      case None  => 
+    }
+    extJars
```
Comment (tillrohrmann) [inline flink-staging/flink-scala-shell/src/main/scala/org.apache.flink/api/scala/FlinkILoop.scala]: Why not doing simply:
```
def getExternalJars(): Array[String] = externalJars.getOrElse(Array.empty[String])
```
?

File: flink-staging/flink-scala-shell/src/main/scala/org.apache.flink/api/scala/FlinkShell.scala [hunk 2]
```diff
@@ -30,7 +31,8 @@ object FlinkShell {
 
     // scopt, command line arguments
     case class Config(port: Int = -1,
```
Comment (tillrohrmann) [inline flink-staging/flink-scala-shell/src/main/scala/org.apache.flink/api/scala/FlinkShell.scala]: parameters should start in a new line with 4 spaces indentation

File: flink-staging/flink-scala-shell/src/main/scala/org.apache.flink/api/scala/FlinkShell.scala [hunk 3]
```diff
@@ -44,6 +46,16 @@ object FlinkShell {
           c.copy (host = x)
       }  text("host specifies host name of running JobManager")
 
+      opt[(String)] ('a',"addclasspath") action {
+        case (x,c) =>
+          val xArray = x.split(":")
+          var extJarArray = new Array[String](xArray.length)
+          for(i <- 0 to xArray.length - 1){
+            extJarArray(i) = xArray(i)
+          }
```
Comment (tillrohrmann) [inline flink-staging/flink-scala-shell/src/main/scala/org.apache.flink/api/scala/FlinkShell.scala]: Why do you copy `xArray` here? And why not simply calling `xArray.clone`?

File: flink-staging/flink-scala-shell/src/main/scala/org.apache.flink/api/scala/FlinkShell.scala [hunk 4]
```diff
@@ -51,14 +63,16 @@ object FlinkShell {
...
-        startShell(config.host,config.port)
+        startShell(config.host,config.port,config.externalJars)
 
       case _ => println("Could not parse program arguments")
     }
   }
 
 
-  def startShell(userHost : String, userPort : Int): Unit ={
+  def startShell(userHost : String,
```
Comment (tillrohrmann) [inline flink-staging/flink-scala-shell/src/main/scala/org.apache.flink/api/scala/FlinkShell.scala]: parameters start in a new line with 4 spaces indentation

**Comment by tillrohrmann** [comment]
Looks good @nikste. I only had some minor comments.

Did you test the code on a cluster? Can we maybe add a test case which tests that external jars are shipped?

File: flink-staging/flink-scala-shell/src/main/scala/org.apache.flink/api/scala/FlinkILoop.scala (hunk 1 above)
Comment (nikste) [inline flink-staging/flink-scala-shell/src/main/scala/org.apache.flink/api/scala/FlinkILoop.scala]: This is much nicer, thanks!

File: flink-staging/flink-scala-shell/src/main/scala/org.apache.flink/api/scala/FlinkShell.scala (hunk 2 above)
Comment (nikste) [inline flink-staging/flink-scala-shell/src/main/scala/org.apache.flink/api/scala/FlinkShell.scala]: This goes totally against my sense of aesthetics, but ok.

File: flink-staging/flink-scala-shell/src/main/scala/org.apache.flink/api/scala/FlinkShell.scala (hunk 3 above)
Comment (nikste) [inline flink-staging/flink-scala-shell/src/main/scala/org.apache.flink/api/scala/FlinkShell.scala]: good point, it's a remainder of an older version with a slightly different approach.

**Comment by nikste** [comment]
Changed according to comments, 
@tillrohrmann I did not check it on a cluster, since I don't have one available at the moment. I checked with a local jobmanager (using "start-local.sh"), that seemed to work. Maybe you can check if it works with a remote cluster as well?

Concerning automated tests, its somewhat tricky. The jar has to be sent to the remote execution and at the same time be available in the shell. You cant add jars from inside the shell (scala 2.10 bug:https://issues.scala-lang.org/browse/SI-6502), only at startup with an additional parameter (see bash script: start-scala-shell.sh).
If I try to invoke the start-scala-shell.sh script from inside java/scala and redirect the std out I only get half of the output (everything till the actual scala shell starts).

Is it possible to add an additional classpath in maven for the tests? Then at least I could write a test, that checks if the jars are shipped to the remote environment..
Or do you have another suggestion maybe?

**Comment by StephanEwen** [comment]
To add more jars to the Classpath, you can write the test in the following way: The test itself is just a wrapper that instantiates a custom classloader with your extra class files. It will the load the test class with Class.forName(classname, true, createdClassLoader).

That way, the test has the extra classes available...

**Comment by nikste** [comment]
This did not work unfortunately, the class was available in the test, but unfortunately not in the shell which is invoked in the test. 
However, if you add the classpath of the external class to ```settings.classpath.value``` of the scala shell before starting it, it seems to work.

I added a test for instantiating and printing a DenseVector with flink-ml jar. This should check if the external jar is sent to the cluster.
The only remaining problem is the name of the jar, which will change if the flink-version changes.

**Comment by nikste** [comment]
The IT case now looks for a jar inside the ml-folder which does not contain "test" or "original".
Thus the IT case should work for different flink-versions as long as the ml module is made before the scala-shell.

**Comment by nikste** [comment]
Could someone please check this, it has been here for a while and I've got some other stuff building on top of this!

**Comment by StephanEwen** [comment]
Sorry, it fell behind in the long list of Pull Requests.
I'll check it and will merge it if Travis gives a green light.

**Pull request closed without merging**